src/Icicle/Avalanche/Statement/Simp/Constructor.hs

```diff
@@ -96,28 +127,51 @@ constructor a_fresh statements
    , Just (_,_,_,_,b) <- fromSum x'
    = b
 
-   | Just (PrimUnsafe (PrimUnsafeArrayIndex _), [XVar _ n, ix]) <- takePrimApps x
-   , Just x' <- get env n
-   , Just (ta, tb, a, b) <- fromZippedArray x'
+   -- update
+   | Just (PrimUpdate (PrimUpdateArrayPut _), [arr,ix,v]) <- takePrimApps x
+   , Just arrx              <- resolve env arr
+   , Just (ta, tb, i, a, b) <- fromSummedArray arrx
+   , Just vx                <- resolve env v
+   , Just (_,_,boool,l,r)   <- fromSum vx
+   , tai                    <- ArrayT BoolT
+   , taa                    <- ArrayT ta
+   , tab                    <- ArrayT tb
+   = case boool of
+      XValue _ _ (VBool False)
+       -> xPrim (PrimMinimal (Min.PrimPair (Min.PrimPairSnd tai taa)))
+          `xApp` (xPrim (PrimUpdate (PrimUpdateArrayPut2 (ArrayT BoolT) (ArrayT ta)))
+                  `xApp` i `xApp` a `xApp` ix `xApp` xValue BoolT (VBool False) `xApp` l)
+      XValue _ _ (VBool True)
+       -> xPrim (PrimMinimal (Min.PrimPair (Min.PrimPairSnd tai tab)))
+          `xApp` (xPrim (PrimUpdate (PrimUpdateArrayPut2 (ArrayT BoolT) (ArrayT tb)))
+                  `xApp` i `xApp` b `xApp` ix `xApp` xValue BoolT (VBool True) `xApp` r)
+      _ -> x
+
+   | Just (PrimUpdate (PrimUpdateArrayPut _), [arr,ix,v]) <- takePrimApps x
+   , Just arrx           <- resolve env arr
+   , Just (ta, tb, a, b) <- fromZippedArray arrx
+   , Just vx             <- resolve env v
```

Comment:
instead of looking in the environment for v, I think you just want to put ``fst v`` and ``snd v``. if ``resolve`` or ``fromPair`` we still want to do this